database/skylink.go

```diff
@@ -0,0 +1,125 @@
+	SkylinkStatusNew = "new"
+	// SkylinkStatusScanning is the status of the skylink while it's being
+	// scanned.
+	SkylinkStatusScanning = "scanning"
+	// SkylinkStatusComplete is the status of the skylink after it's scanned.
+	SkylinkStatusComplete = "complete"
+)
+
+// Skylink represents a skylink in the queue and holds its scanning status.
+//
+// ClamAV typically limits the amount of data it scans, e.g. it would only scan
+// the first 16MiB of data in a given file. ScannedAllContent marks if we've
+// managed to scan all the content or just its beginning.
+//
+// Multiple skylinks can point to the same merkle root but use different offsets
+// on the data. Since we're blocking the entire merkle root, we should be
+// scanning all possible (for the size of the data) offsets. ScannedAllOffsets
+// marks if we have done that or not.
+//
+// Timestamp marks the last status change that happened to the record. It
+// can be the time when it was created, locked for scanning, or scanned.
+type Skylink struct {
+	ID                   primitive.ObjectID `bson:"_id,omitempty" json:"-"`
+	Hash                 crypto.Hash        `bson:"hash" json:"hash"`
+	Skylink              string             `bson:"skylink" json:"skylink"`
+	Status               string             `bson:"status" json:"status"`
+	Infected             bool               `bson:"infected" json:"infected"`
+	InfectionDescription string             `bson:"infection_description" json:"infectionDescription"`
+	ScannedAllContent    bool               `bson:"scanned_all_content" json:"scannedAllContent"`
+	ScannedAllOffsets    bool               `bson:"scanned_all_offsets" json:"scannedAllOffsets"`
+	Size                 uint64             `bson:"size" json:"size"`
+	Timestamp            time.Time          `bson:"timestamp" json:"timestamp"`
+}
+
+// LoadString parses a skylink from string and populates all required fields.
+func (s *Skylink) LoadString(skylink, portal string) error {
+	if !accdb.ValidSkylinkHash(skylink) {
+		return ErrInvalidSkylink
+	}
+	s.Skylink = skylink
+	var sl skymodules.Skylink
+	err := sl.LoadString(skylink)
+	if err != nil {
+		return errors.AddContext(err, ErrInvalidSkylink.Error())
+	}
+
+	switch {
+	case sl.IsSkylinkV1():
+		s.Hash = crypto.HashObject(sl.MerkleRoot())
+	case sl.IsSkylinkV2():
+		slv1, err := resolveSkylinkV2(sl, portal)
+		if err != nil {
+			return errors.AddContext(err, "unable to resolve v2 skylink")
+		}
+		s.Hash = crypto.HashObject(slv1.MerkleRoot())
+	default:
+		return renter.ErrInvalidSkylinkVersion
+	}
+
+	if s.Timestamp.IsZero() {
+		s.Timestamp = time.Now().UTC()
+	}
+	if s.Status == "" {
+		s.Status = SkylinkStatusNew
+	}
+	return nil
+}
+
+// resolveSkylinkV2 returns the v1 skylink to which the given v2 skylink is
+// currently pointing. Resolves up to three levels of nested v2 skylinks.
+func resolveSkylinkV2(s skymodules.Skylink, portal string) (*skymodules.Skylink, error) {
```

Comment:
If we put the requirement on the caller we can avoid this code here. If we increase the recursive depth of the v2 skylink in the future we will need to remember to update this too. That is probably a good reason to just restrict the scanner to only accept v1 skylinks. The portal should have the v1 skylink anyways by the end of the request. And the scanner won't be scanning before the request completes anyways so we don't lose anything by waiting at the portal level. 

The skylink would only not be added if the portal crashed/exited before the request is complete which would mean the user didn't get the full request anyways.

Reply:
fine with a follow up

Reply:
F/U issue for discussion: https://github.com/SkynetLabs/malware-scanner/issues/5